src/drivers/rest/data.rs

```diff
@@ -0,0 +1,164 @@
+    };
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct MAVLinkMessage<T> {
+    pub header: mavlink::MavHeader,
+    pub message: T,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize)]
+pub struct MAVLinkVehiclesData {
+    vehicles: HashMap<u8, MAVLinkVehicleData>,
+}
+
+impl MAVLinkVehiclesData {
+    fn update(&mut self, message: MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let vehicle_id = message.header.system_id;
+        self.vehicles
+            .entry(vehicle_id)
+            .or_insert(MAVLinkVehicleData {
+                id: vehicle_id,
+                components: HashMap::new(),
+            })
+            .update(&message);
+    }
+
+    pub fn pointer(&self, path: &str) -> String {
+        if path.is_empty() {
+            return serde_json::to_string_pretty(self).unwrap();
+        }
+
+        let path = format!("/{path}");
+
+        dbg!(&path);
+
+        if path == "/vehicles" {
+            return serde_json::to_string_pretty(&self.vehicles).unwrap();
+        };
+
+        let value = serde_json::to_value(self).unwrap();
+        return match value.pointer(&path) {
+            Some(content) => serde_json::to_string_pretty(content).unwrap(),
+            None => "None".into(),
+        };
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleData {
+    id: u8,
+    components: HashMap<u8, MAVLinkVehicleComponentData>,
+}
+
+impl MAVLinkVehicleData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let component_id = message.header.component_id;
+        self.components
+            .entry(component_id)
+            .or_insert(MAVLinkVehicleComponentData {
+                id: component_id,
+                messages: HashMap::new(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkVehicleComponentData {
+    id: u8,
+    messages: HashMap<String, MAVLinkMessageStatus>,
+}
+
+impl MAVLinkVehicleComponentData {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        let message_name = message.message.message_name().to_string();
+        self.messages
+            .entry(message_name)
+            .or_insert(MAVLinkMessageStatus {
+                message: message.message.clone(),
+                status: Status::default(),
+            })
+            .update(message);
+    }
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+struct MAVLinkMessageStatus {
+    message: mavlink::ardupilotmega::MavMessage,
+    status: Status,
+}
+
+impl MAVLinkMessageStatus {
+    fn update(&mut self, message: &MAVLinkMessage<mavlink::ardupilotmega::MavMessage>) {
+        self.message = message.message.clone();
+        self.status.update();
+    }
+}
+
+#[derive(Debug)]
+struct Data {
+    messages: Arc<Mutex<MAVLinkVehiclesData>>,
+}
+
+#[derive(Default, Clone, Debug, Deserialize, Serialize)]
```

Comment:
Because we dropped clone in MAVLinkMessageStatus, we don't need to impl clone here:

```suggestion
#[derive(Default, Clone, Debug, Deserialize, Serialize)]
```

Reply:
done